File: app/src/main/java/com/crossbowffs/quotelock/backup/LocalBackup.java
package com.crossbowffs.quotelock.backup;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Environment;

import androidx.core.app.ActivityCompat;

import com.crossbowffs.quotelock.R;
import com.crossbowffs.quotelock.utils.AppExecutors;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.OutputStream;

/**
 * Reference: <a href="https://github.com/prof18/Database-Backup-Restore/blob/master/app/src/main/java/com/prof/dbtest/backup/LocalBackup.java">Database-Backup-Restore</a>
 *
 * @author Yubyf
 */
public class LocalBackup {

    private static final String[] PERMISSIONS_STORAGE = {
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };
    public static final int REQUEST_CODE_PERMISSIONS_BACKUP = 55;
    public static final int REQUEST_CODE_PERMISSIONS_RESTORE = 43;

    /** check permissions. */
    public static boolean verifyPermissions(Activity activity, int requestCode) {
        // Check if we have read or write permission
        int writePermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE);
        int readPermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE);

        if (writePermission != PackageManager.PERMISSION_GRANTED || readPermission != PackageManager.PERMISSION_GRANTED) {
            // We don't have permission so prompt the user
            ActivityCompat.requestPermissions(
                    activity,
                    PERMISSIONS_STORAGE,
                    requestCode
            );
            return false;
        }
        return true;
    }

    public static void handleRequestPermissionsResult(int[] grantResults, ProgressCallback callback,
                                                      Runnable action) {
        if (grantResults.length < 2 || grantResults[0] != PackageManager.PERMISSION_GRANTED
                || grantResults[1] != PackageManager.PERMISSION_GRANTED) {
            callback.failure("Please grant external storage permission and retry.");
            return;
        }
        action.run();
    }

    /** ask to the user a name for the backup and perform it. The backup will be saved to a custom folder. */
    public static void performBackup(Activity activity, String databaseName, ProgressCallback callback) {
        if (!verifyPermissions(activity, REQUEST_CODE_PERMISSIONS_BACKUP)) {
            callback.failure("Please grant external storage permission and retry.");
            return;
        }

        AppExecutors.getInstance().diskIO().execute(() -> {
            File folder = new File(Environment.getExternalStorageDirectory() + File.separator + activity.getResources().getString(R.string.quotelock));

            boolean success = true;
            if (!folder.exists()) {
                success = folder.mkdirs();
            }
            if (success) {
                String out = folder.getAbsolutePath() + File.separator + databaseName;
                try {
                    backup(activity, databaseName, out);
                    callback.safeSuccess();
                } catch (Exception e) {
                    e.printStackTrace();
                    callback.safeFailure(e.getMessage());
                }
            } else {
                callback.safeFailure("Unable to create directory. Retry");
            }
        });
    }

    /** ask to the user what backup to restore */
    public static void performRestore(Activity activity, String databaseName, ProgressCallback callback) {
        if (!verifyPermissions(activity, REQUEST_CODE_PERMISSIONS_RESTORE)) {
            callback.failure("Please grant external storage permission and retry.");
        }

        AppExecutors.getInstance().diskIO().execute(() -> {
            File folder = new File(Environment.getExternalStorageDirectory() + File.separator + activity.getResources().getString(R.string.quotelock));
            if (folder.exists()) {
                File file = new File(folder, databaseName);
                if (!file.exists()) {
                    callback.safeFailure("Backup file not exists.\nDo a backup before a restore!");
                    return;
                }
                try {
                    importDb(activity, databaseName, file.getAbsolutePath());
                    callback.safeSuccess();
                } catch (Exception e) {
                    e.printStackTrace();
                    callback.safeFailure(e.getMessage());
                }
            } else {
                callback.safeFailure("Backup folder not present.\nDo a backup before a restore!");
            }
        });
    }

    private static boolean backup(Context context, String databaseName, String outFileName) throws Exception {
        //database path
        final String inFileName = context.getDatabasePath(databaseName).toString();

        try {

            File dbFile = new File(inFileName);
            FileInputStream fis = new FileInputStream(dbFile);

            // Open the empty db as the output stream
            OutputStream output = new FileOutputStream(outFileName);

            // Transfer bytes from the input file to the output file
            byte[] buffer = new byte[1024];
            int length;
            while ((length = fis.read(buffer)) > 0) {
                output.write(buffer, 0, length);
            }

            // Close the streams
            output.flush();
            output.close();
            fis.close();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception("Unable to backup database. Retry");
        }
    }

    private static boolean importDb(Context context, String databaseName, String inFileName) throws Exception {
        final String outFileName = context.getDatabasePath(databaseName).toString();

        try {

            File dbFile = new File(inFileName);
            FileInputStream fis = new FileInputStream(dbFile);

            // Open the empty db as the output stream
            OutputStream output = new FileOutputStream(outFileName);

            // Transfer bytes from the input file to the output file
            byte[] buffer = new byte[1024];
            int length;
            while ((length = fis.read(buffer)) > 0) {
                output.write(buffer, 0, length);
            }

            // Close the streams
            output.flush();
            output.close();
            fis.close();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception("Unable to import database. Retry");
        }
    }
}

File: app/src/main/java/com/crossbowffs/quotelock/backup/LocalBackup.kt
package com.crossbowffs.quotelock.backup

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.ContentValues
import android.content.Context
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import androidx.core.app.ActivityCompat
import com.crossbowffs.quotelock.app.App
import com.crossbowffs.quotelock.collections.database.QuoteCollectionContract
import com.crossbowffs.quotelock.collections.database.QuoteCollectionDatabase
import com.crossbowffs.quotelock.collections.database.quoteCollectionDatabase
import com.crossbowffs.quotelock.utils.fromFile
import com.crossbowffs.quotelock.utils.ioScope
import com.crossbowffs.quotelock.utils.toFile
import com.yubyf.quotelockx.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

/**
 * Reference: [Database-Backup-Restore](https://github.com/prof18/Database-Backup-Restore/blob/master/app/src/main/java/com/prof/dbtest/backup/LocalBackup.java)
 *
 * @author Yubyf
 */
object LocalBackup {
    private val PERMISSIONS_STORAGE = arrayOf(
        Manifest.permission.WRITE_EXTERNAL_STORAGE
    )
    const val REQUEST_CODE_PERMISSIONS_BACKUP = 55
    const val REQUEST_CODE_PICK_FILE = 43

    val PREF_BACKUP_ROOT_DIR: File =
        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
    val PREF_BACKUP_RELATIVE_PATH = App.INSTANCE.resources.getString(R.string.quotelockx)

    /** Check necessary permissions.  */
    private fun verifyPermissions(activity: Activity, requestCode: Int): Boolean {
        // Check if we have write permission
        return if (ActivityCompat.checkSelfPermission(activity,
                Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
        ) {
            // We don't have permission so prompt the user
            ActivityCompat.requestPermissions(
                activity,
                PERMISSIONS_STORAGE,
                requestCode
            )
            false
        } else true
    }

    fun handleRequestPermissionsResult(
        grantResults: IntArray, callback: ProgressCallback,
        action: Runnable,
    ) {
        if (grantResults.size < 2 || grantResults[0] != PackageManager.PERMISSION_GRANTED
            || grantResults[1] != PackageManager.PERMISSION_GRANTED
        ) {
            callback.failure("Please grant external storage permission and retry.")
            return
        }
        action.run()
    }

    /** Ask the user a name for the backup and perform it.
     *  The backup will be saved to a custom folder in [Environment.DIRECTORY_DOWNLOADS].
     */
    fun performBackup(activity: Activity, databaseName: String, callback: ProgressCallback) {
        // Use MediaStore to save files in public directories above Android Q.
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q
            && !verifyPermissions(activity, REQUEST_CODE_PERMISSIONS_BACKUP)
        ) {
            callback.failure("Please grant external storage permission and retry.")
            return
        }
        ioScope.launch {
            try {
                backupDb(activity, databaseName)
                callback.safeSuccess()
            } catch (e: Exception) {
                e.printStackTrace()
                callback.safeFailure(e.message)
            }
        }
    }

    /** Ask the user which backup to restore from. */
    fun performRestore(
        activity: Activity,
        databaseName: String,
        pickedUri: Uri,
        callback: ProgressCallback,
    ) {
        ioScope.launch scope@{
            try {
                importDb(activity, databaseName, pickedUri)
                callback.safeSuccess()
            } catch (e: Exception) {
                e.printStackTrace()
                callback.safeFailure(e.message)
            }
        }
    }

    @Throws(Exception::class)
    private fun backupDb(context: Context, databaseName: String): Boolean {
        //database path
        val inFileName = context.getDatabasePath(databaseName).toString()
        return try {
            val dbFile = File(inFileName)
            copyFileToDownloads(context, dbFile) != null
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception("Unable to backup database. Retry")
        }
    }

    @Suppress("BlockingMethodInNonBlockingContext")
    @Throws(Exception::class)
    private suspend fun importDb(context: Context, databaseName: String, fileUri: Uri): Boolean {
        val temporaryDatabaseFile = File(context.cacheDir, databaseName)
        return try {
            context.contentResolver.openInputStream(fileUri)?.toFile(temporaryDatabaseFile)
                ?: throw Exception()
            importCollectionDatabaseFrom(context, temporaryDatabaseFile)
            true
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception("Unable to import database. Retry")
        }
    }

    @SuppressLint("Range")
    private fun copyFileToDownloads(context: Context, file: File): Uri? {
        // Generate export name with timestamp
        val date = Date(System.currentTimeMillis())
        val simpleDateFormat = SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault())
        val exportName = file.name.replace(file.nameWithoutExtension,
            file.nameWithoutExtension.plus("_").plus(simpleDateFormat.format(date)))

        // Copy file
        val resolver = context.contentResolver
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            // Adapt the scope storage above Android Q by MediaStore.
            val contentValues = ContentValues().apply {
                put(MediaStore.MediaColumns.RELATIVE_PATH,
                    Environment.DIRECTORY_DOWNLOADS + File.separatorChar + PREF_BACKUP_RELATIVE_PATH)
                put(MediaStore.MediaColumns.DISPLAY_NAME, exportName)
                put(MediaStore.MediaColumns.MIME_TYPE, "application/vnd.sqlite3")
            }
            resolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, contentValues)
        } else {
            // Save file through File API on Android P-.
            val targetFile = File(File(PREF_BACKUP_ROOT_DIR, PREF_BACKUP_RELATIVE_PATH), exportName)
            targetFile.parentFile?.mkdirs()
            Uri.fromFile(targetFile)
        }?.also { resolver.openOutputStream(it)?.fromFile(file.absoluteFile) }
    }
}

/**
 * Import collection database from .db file by replacing data contents.
 */
@Throws(NoSuchElementException::class)
suspend fun importCollectionDatabaseFrom(context: Context, file: File) {
    withContext(Dispatchers.IO) {
        val temporaryCollectionDatabase =
            QuoteCollectionDatabase.openTemporaryDatabaseFrom(context,
                "${QuoteCollectionContract.DATABASE_NAME}_${System.currentTimeMillis()}", file)
        val collections = temporaryCollectionDatabase.dao().getAll().first()
        quoteCollectionDatabase.dao().clear()
        if (collections.isNotEmpty()) {
            quoteCollectionDatabase.dao().insert(collections)
        }
        temporaryCollectionDatabase.close()
    }
}